### DISCUSSION TITLE
Concurrency check cancels job if using ${{github.job}} and Job Name contains '&'

### DISCUSSION THREAD

**Discussion by bruschke** [community/community#13496] [Actions] [answered]
Discovered the concurrency functionality recently (which is amazing, thank you so much).  However, I think I found a small bug.  A couple of our jobs kept being cancelled and showing this reason:

Canceling since a higher priority waiting request for '-refs/heads/master' exists

`  push_docker_image: `
`    name: Build & Push Docker Image `
`    if: github.ref == 'refs/heads/master' `
`    runs-on: ubuntu-latest `
`    concurrency: `
`      group: ${{ github.job }}-${{ github.ref }}  `
`      cancel-in-progress: true `
`    steps:`

As you can see in our code the group was set to '${{ github.job }}-${{ github.ref }} but the github.job portion was coming back empty.  After hard coding the group to be like the following, everything worked.  I noticed that the jobs that were consistently failing were the jobs that had an '&' in the name.

Work around (hard code the name of the job in the group):
`    concurrency: `
`      group: build-push-docker-${{ github.ref }} `
`      cancel-in-progress: true `

Thanks again for the great work and the recent updates.  Love the ability to set concurrency combined with rerun failed jobs.

**Comment by chrispat** [comment, marked as answer]
Thank you for the feedback.  

I will file an issue around the usage of & in the concurrency key.  The `${{ github.job }}` being empty may be harder to fix as no job really exists when the replacement is done in this case.  I will take a look and see if we can clarify this in the documentation.

**Reply by myitcv** [reply]
We appear to be running into something similar, @chrispat. 

Is there an issue/some docs somewhere I can read to get up to speed on the problem and solution?

Edit: we appear to be seeing a different problem.

**Comment by brcrista** [comment]
Thanks for reporting!  After investigating, I wasn't able to repro the issue with `&` in the job name.  It seems like the issue is just the expression expansion of the concurrency group:

```yml
 group: ${{ github.job }}-${{ github.ref }} 
```

The value of `group` is evaluated on the GitHub Actions server, not on the runner.  However, the `github.job` context is set by the runner, so it's empty here ([docs](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context)):

> The [job_id](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_id) of the current job.
> Note: This context property is set by the Actions runner, and is only available within the execution steps of a job. Otherwise, the value of this property will be null.

That's why you get the message:

```
Canceling since a higher priority waiting request for '-refs/heads/master' exists
```

**Reply by quantizor** [reply]
I'm actually seeing this exact issue on a private repository but without the race condition issue. For example, the concurrency key is "i18n-download-4066", nothing missing.

![Screenshot 2024-01-25 at 4 43 56 PM](https://github.com/community/community/assets/570070/2bc1c312-c7a9-4793-8b3a-c67554334155)

![Screenshot 2024-01-25 at 4 49 54 PM](https://github.com/community/community/assets/570070/e779ea3f-b771-485a-bb39-16185d1bbe1a)


Two events came in at the same time and the later one should have "won" but didn't with this error annotation:

```
Canceling since a higher priority waiting request for 'i18n-download-4066' exists
```

**Reply by ericksprengel** [reply]
@quantizor same issue here.

The executions were dispatched at the same time, but the second was canceled with:
```
Canceling since a higher priority waiting request for 'Tests-7552' exists
```
And the side-effect was that the PR got stuck with `Tests` workflow in `queued` status. I forced a re-run on the cancelled execution and it worked, but it's just a workaround.

![image](https://github.com/community/community/assets/1028164/6bfcfe1c-ec48-4e59-8df8-18857b659cb7)

**Reply by JD10NN3** [reply]
Same here. 
Any resolution for this issue ?

**Reply by edoardocavazza** [reply]
Has anyone found a solution?

**Reply by amberto-qe** [reply]
I'm in the same issue, it cancel though i set
<img width="244" height="63" alt="image" src="https://github.com/user-attachments/assets/097247d2-c9a3-40bd-b34e-9b6e47013f09" />

**Reply by javier-ruiz-voxsmart** [reply]
This problem is still happening.
<img width="934" height="74" alt="image" src="https://github.com/user-attachments/assets/5b2a21f6-4033-4857-84c4-9f17d9373f6e" />

**Reply by amberto-qe** [reply]
Hi @chrispat, @brcrista, as using Github Action with these setting, we still encounter the error that causing accidentally cancel job as above. Can you help investigating these issue?
